packages/worker/src/worker-tuner.ts

```diff
@@ -73,15 +87,127 @@ export interface ResourceBasedTuner {
 }
 
-/**
- * This tuner allows for different slot suppliers for different slot types.
- *
- * @experimental
- */
-export interface TunerHolder {
-  workflowTaskSlotSupplier: SlotSupplier;
-  activityTaskSlotSupplier: SlotSupplier;
-  localActivityTaskSlotSupplier: SlotSupplier;
+export type SlotInfo = WorkflowSlotInfo | ActivitySlotInfo | LocalActivitySlotInfo;
+
+export interface WorkflowSlotInfo {
+  type: 'workflow';
+  workflowId: string;
+  runId: string;
+}
+
+export interface ActivitySlotInfo {
+  type: 'activity';
+  activityId: string;
+}
+
+export interface LocalActivitySlotInfo {
+  type: 'local-activity';
+  activityId: string;
+}
+
+export interface UserSlotSupplier<SI extends SlotInfo> {
+  type: 'user';
+
+  /**
+   * This function is called before polling for new tasks. Your implementation should block until a
+   * slot is available then return a permit to use that slot.
+   *
+   * // TODO: How to handle cancellation? AbortController? CancelToken?
+   *
+   * @param ctx The context for slot reservation.
+   * @returns A permit to use the slot which may be populated with your own data.
+   */
+  reserveSlot(ctx: SlotReserveContext): Promise<SlotPermit>;
+
+  /**
+   * This function is called when trying to reserve slots for "eager" workflow and activity tasks.
+   * Eager tasks are those which are returned as a result of completing a workflow task, rather than
+   * from polling. Your implementation must not block, and If a slot is available, return a permit
+   * to use that slot.
+   *
+   * @param ctx The context for slot reservation.
+   * @returns Maybe a permit to use the slot which may be populated with your own data.
+   */
+  tryReserveSlot(ctx: SlotReserveContext): SlotPermit | undefined;
+
+  /**
+   * This function is called once a slot is actually being used to process some task, which may be
+   * some time after the slot was reserved originally. For example, if there is no work for a
+   * worker, a number of slots equal to the number of active pollers may already be reserved, but
+   * none of them are being used yet. This call should be non-blocking.
+   *
+   * @param ctx The context for marking a slot as used.
+   */
+  markSlotUsed(slot: SlotMarkUsedContext<SI>): void;
+
+  /**
+   * This function is called once a permit is no longer needed. This could be because the task has
+   * finished, whether successfully or not, or because the slot was no longer needed (ex: the number
+   * of active pollers decreased). This call should be non-blocking.
+   *
+   * @param ctx The context for releasing a slot.
+   */
+  releaseSlot(slot: SlotReleaseContext<SI>): void;
+}
+
+export interface SlotPermit {
+  // Use this field to associate any data you like with a particular slot
+  data?: any;
+}
+
+export interface SlotReserveContext {
+  // The name of the task queue for which this reservation request is associated
+  taskQueue: string;
+  // The build id of the worker that is requesting the reservation
```

Comment:
Docs for `workerIdentity` and `workerBuildId` are reversed.

Also, make all of these `/** ... */` comments, so they would be visible in API doc.

Reply:
Thanks. I thought // still counted for API docs. Will fix